### PR TITLE
Remove providerConfig from NetworkController

### LINKED ISSUE #4185: Remove providerConfig from NetworkController
Historically, the `providerConfig` property in NetworkController has been used to switch and track the currently selected network as well as provide access to information about that network. That is now achieved with a combination of the `selectedNetworkClientId` and `networkConfigurations` properties as well as the `configuration` property on the NetworkClient interface. This means that we no longer need `providerConfig` and we can remove this redundant state.

Removing `providerConfig` is one thing, but there are also a ton of packages which make use of this property to access the currently selected chain, and we would want to make use of `selectedNetworkClientId` in combination with the `NetworkController:getNetworkClientById` action to do this instead.

## Acceptance Criteria

- [x] `providerConfig` is removed from NetworkController (replaced with `selectedNetworkClientId` and/or `networkConfigurations); NetworkController tests pass
- [x] [assets-controllers] `AccountTrackerController` is updated to follow suit and tests pass
- [x] [assets-controllers] `AssetsContractController` is updated to follow suit and tests pass
- [x] [assets-controllers] `NftController` is updated to follow suit and tests pass
- [x] [assets-controllers] `NftDetectionController` is updated to follow suit and tests pass
- [x] [assets-controllers] `TokensController` is updated to follow suit and tests pass
- [x] [assets-controllers] `TokenListController` is updated to follow suit and tests pass
- [x] [assets-controllers] `TokenRatesController` is updated to follow suit and tests pass
- [x] `EnsController` is updated to follow suit and tests pass
- [x] `GasFeeController` is updated to follow suit and tests pass
- [x] `SelectedNetworkController` tests are updated to follow suit and tests pass
- [x] `TransactionController` is updated to follow suit and tests pass

### PR DESCRIPTION
**Pull request by mcmire**
## Explanation

Historically, the `providerConfig` property in NetworkController has been used to track the currently selected network as well as provide access to information about that network. The selected network is now tracked via `selectedNetworkClientId`, and information about that network can be retrieved by looking at the `networkConfigurations` property or the `configuration` property on the NetworkClient interface. This means that we no longer need `providerConfig` and we can remove this redundant state.

## References

Fixes #4185.

## Changelog

(Updated in PR)

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/assets-controllers/src/AssetsContractController.ts [hunk 1]
```diff
@@ -154,10 +154,15 @@ export class AssetsContractController extends BaseControllerV1<
...
 
-    onNetworkDidChange((networkState) => {
-      if (this.config.chainId !== networkState.providerConfig.chainId) {
+    onNetworkDidChange(({ selectedNetworkClientId }) => {
+      const selectedNetworkClient = getNetworkClientById(
+        selectedNetworkClientId,
+      );
+      const { chainId } = selectedNetworkClient.configuration;
+
+      if (this.config.chainId !== chainId) {
```
Comment (legobeat) [inline packages/assets-controllers/src/AssetsContractController.ts:163]: Is there a test-case for when `chainId` differs? If not, that could explain the reduced coverage %?

Comment (mcmire) [reply packages/assets-controllers/src/AssetsContractController.ts:163]: I looked into this and I don't think it's due to additional logic. Somehow I have made something that was inadvertently tested not tested anymore. Regrettably I have added an `istanbul` directive to ignore the logic in question: https://github.com/MetaMask/core/pull/4254/commits/be7f9c2b55b8ff0ef12dc9004196cc48b74e3dbc

**Comment by mcmire** [comment]
I'm going to break this up into multiple PRs as this one is rather large and difficult to review. Placing in draft until that's complete.

File: packages/assets-controllers/src/AssetsContractController.ts (hunk 1 above)
Comment (mcmire) [reply packages/assets-controllers/src/AssetsContractController.ts]: This change disappeared when this PR was broken up into multiple PRs.

**Review by mikesposito — APPROVED** [review]

**Comment by mcmire** [comment]
Conflicts resolved.

**Comment by mcmire** [comment]
Okay! Tests should pass again.

File: packages/network-controller/tests/NetworkController.test.ts [hunk 2]
```diff
@@ -29,11 +29,14 @@ import type {
...
 import { NetworkController } from '../src/NetworkController';
-import type { Provider } from '../src/types';
+import type { NetworkClientConfiguration, Provider } from '../src/types';
 import { NetworkClientType } from '../src/types';
+import {
+  buildCustomNetworkClientConfiguration,
+  buildInfuraNetworkClientConfiguration,
+} from './helpers';
 
 jest.mock('../src/create-network-client');
```
[the remark this answers is not in the document]
Comment (mcmire) [reply packages/network-controller/tests/NetworkController.test.ts:42]: Good question. This change was made by another team mistakenly and wasn't caught, and has now persisted. I agree that we should move this to `src/` however.

File: packages/network-controller/src/NetworkController.ts [hunk 3]
```diff
@@ -1569,106 +1381,61 @@ export class NetworkController extends BaseController<
...
-    let networkClientId: NetworkClientId;
-    if (isInfuraProviderConfig(providerConfig)) {
-      const networkClientType = NetworkClientType.Infura;
-      networkClientId = buildInfuraNetworkClientId(providerConfig);
-      const builtInNetworkClientRegistry =
-        this.#autoManagedNetworkClientRegistry[networkClientType];
-      autoManagedNetworkClient =
-        builtInNetworkClientRegistry[networkClientId as BuiltInNetworkClientId];
-      if (!autoManagedNetworkClient) {
+      // This is impossible to reach
```
[the remark this answers is not in the document]
Comment (mcmire) [reply packages/network-controller/src/NetworkController.ts:1413]: The test we'd need to write here is not what happens when `isInfuraNetworkType` returns true or false, but what happens when `isInfuraNetworkType` returns true and yet `possibleAutoManagedNetworkClient` is `undefined` (which means there is no network client registered for the `networkClientId` when it refers to an Infura network).

The reason why this is impossible is that `autoManagedNetworkClientRegistry[NetworkClientType.Infura]` is automatically populated with all of the Infura networks that we know about, and there is no way to stop that from happening. And `isInfuraNetworkType` checks the given `networkClientId` against the same list of known Infura networks.

**Pull request merged**